### PR TITLE
Add the ability to register additional key handlers.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

Adding a way to handle key events outside of the Focus system.

Other than setting a single callback, there isn't a way to get key events and handle them if you also want to use the focus system.  This allows both, because the focus system is the "primary" handler of keys, but if any of these new registered handlers returns true, then the event is considered handled.  It will still pass the event to all of the other handlers, but will tell the platform that the event has been handled.

This allows use cases where listeners don't want to necessarily have focus, but they also don't want to double-handle keys that have already been handled by the focus system. It could also be used to disallow text input, since a handler would just have to always return true, and no events would make it to the text input plugins on the platform side.

## Tests

- Added tests for handling key events.

### CODE REVIEW

**Comment by gspencergoog** [comment]
@HansMuller, @dkwingsmt and @goderbauer, let me know what you think of this concept.

**Comment by gspencergoog** [comment]
cc @darrenaustin you too!

**Comment by dkwingsmt** [comment]
So in a nutshell, is this `addListener` but with a return value for handling?

**Comment by gspencergoog** [comment]
> So in a nutshell, is this addListener but with a return value for handling?

Yes, but unlike the listeners, it will skip these new handlers if the focus system handles the key.

That way the focus system is still able to stop propagation of a key if it handles it, but if nothing is focused, you can still handle a key and stop propagation of it in the platform code.

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 1]
```diff
@@ -604,8 +609,38 @@ class RawKeyboard {
...
+  /// Register a handler that is called when the user presses or releases
+  /// a hardware keyboard key if the [keyEventHandler] didn't handle it.
+  ///
+  /// These events will be delivered after the [keyEventHandler] (typically set
+  /// by the [FocusManager]) has returned false. If the [keyEventHandler]
+  /// returns true, then handlers registered in this way will not be called, as
+  /// the key is considered handled already.
+  ///
+  /// If any registered handler returns true, then the key event will be
+  /// considered handled by the platform, but all registered handlers will still
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/raw_keyboard.dart:623]: So when `keyEventHandler` says things are handled, these are not called, but when one of these says things are handled, the other ones are still called (and they will not know that the event was already handled?)? Why the shift in semantics?

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 2]
```diff
@@ -604,8 +609,38 @@ class RawKeyboard {
...
   }
 
-  /// A handler for hardware keyboard events that will stop propagation if the
-  /// handler returns true.
+  final List<RawKeyEventHandler> _keyHandlers = <RawKeyEventHandler>[];
+
+  /// Register a handler that is called when the user presses or releases
+  /// a hardware keyboard key if the [keyEventHandler] didn't handle it.
+  ///
+  /// These events will be delivered after the [keyEventHandler] (typically set
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/raw_keyboard.dart:617]: Do we still need `keyEventHandler` then? Or could the focus system also just use this mechanism?

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 3]
```diff
@@ -604,8 +609,38 @@ class RawKeyboard {
...
+  ///
+  /// If you want to listen to key events, but not handle them, use
+  /// [addListener] instead.
+  ///
+  /// Key handlers can be removed with [removeKeyHandler].
+  void addKeyHandler(RawKeyEventHandler handler) {
+    _keyHandlers.add(handler);
+  }
+
+  /// Stop calling the given handler every time the user presses or releases a
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/raw_keyboard.dart:634]: nit: the handler wasn't called every time a key is pressed, right? Only if `keyEventHandler` didn't handle it, etc.

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 4]
```diff
@@ -604,8 +609,38 @@ class RawKeyboard {
...
+  /// hardware keyboard key.
+  ///
+  /// Key handlers can be added with [addKeyHandler].
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/raw_keyboard.dart:637]: We should also document here and in addKeyHandler what happens if a handler itself removes or adds another handler. Will they get called this iteration? etc.

See doc on https://master-api.flutter.dev/flutter/foundation/ChangeNotifier/addListener.html for example.

**Comment by goderbauer** [comment]
The three different handler types appear to add a lot of complexity. Is it necessary that we have three different ones?

> This allows use cases where listeners don't want to necessarily have focus, but they also don't want to double-handle keys that have already been handled by the focus system. It could also be used to disallow text input, since a handler would just have to always return true, and no events would make it to the text input plugins on the platform side.

Do we have an actual use case for this that we need to enable? How common is this? In theory one could potentially also write some third-party "middleware" that registers itself for keyEventHandler, forwards that call to the focus system, and also gives you the option to register other listeners as is implemented in this PR?

**Comment by goderbauer** [comment]
I wonder if we can replace all this with just one listener type to simplify the API. E.g. a passive listener could just be a `addKeyHandler` that always say "I didn't handle the event" - and could the `keyEventHandler` just be a regular `addKeyHandler` as well?

**Comment by gspencergoog** [comment]
@goderbauer The use case for this is if you wanted to handle keys globally instead of needing to be in the focus chain, but not just listen to them, actually handle them. For instance, if a game doesn't want to deal with focus, but needs to handle the keys so that they don't propagate to the OS. For example, macOS (by design) beeps when an unhandled key is received, and an app could avoid that if they handle all keys that the focus system doesn't. For some apps (mainly games), it's just easier to listen to all the keys all the time, and handle them when you do something with them.

The reason for three different methods is because `addListener` API is legacy, and used extensively: I'd remove it in a heartbeat if I could without breaking the world.

The reason the focus system can't use this new mechanism is because the registration mechanism doesn't really lend itself to ordering. Ideally, the focus system could use it, if the order of the handlers could be controlled.

Maybe this API needs to change to have `addHandlerAtBeginning` and `addHandlerAtEnd`, etc. so that registrants can control whether they get key events early or late, but that probably means we want a full list interface, really, and that complicates the API.

**Comment by dkwingsmt** [comment]
I don't think adding global handler is the way, because the root cause is that the "game app" wants to be the keyboard focus as long as Flutter is focused. This PR rewrites `RawKeyboardListener` into methods of a global singleton, but what about shortcuts? We're going to rewrite `SingleActivator` into methods of `RawKeyboard` (the cookbook already does so), and in the futurewe'll have to do the same to every keyboard widgets.

I think the better way is to find a way to make a focusable widget permanently focused (focused as long as Flutter is focused). Will `FocusManager.rootScope` make it?

**Comment by gspencergoog** [comment]
I don't understand what shortcuts has to do with it.  Shortcuts are inherently tied to focus, and will work regardless of what global handlers there are: they are essentially `Focus` widgets with a fancy `onKey` handler.  This PR doesn't change that.

Yes, the `onKey` handler for `FocusManager.rootScope` would always be called after any other node has a chance to handle it, so you could set that to something that calls a set of registered key handlers instead of using this interface. There's no way to make any widget "permanently focused", and focusing the root scope is a bad idea if you still want to use focus traversal (because it has no context to find what is "next").

If an app isn't interested in focus at all, they can just set `RawKeyboard.instance.keyEventHandler` and cut the focus manager out of the loop entirely.

So, I guess we don't *really* need this interface.  I'll make a different PR that adds documentation to various places describing how to bypass the focus manager, or set an onKey handler on the root scope.

**Pull request closed without merging**